File: src/main/java/com/github/iamniklas/liocore/procedures/variants/FadeToUniformColorProcedure.java
package com.github.iamniklas.liocore.procedures.variants;

import com.github.iamniklas.liocore.config.ProgramConfiguration;
import com.github.iamniklas.liocore.led.LEDDataBundle;
import com.github.iamniklas.liocore.led.colorspace.ColorRGB;
import com.github.iamniklas.liocore.led.colorspace.ColorRGBA;
import com.github.iamniklas.liocore.procedures.Procedure;

public class FadeToUniformColorProcedure extends Procedure {

    private LEDDataBundle bundle;

    private ColorRGB baseColor;
    private ColorRGB targetColor;
    private float duration = 0.0f;
    private int counter = 0;
    private float alphaStep = 0.0f;
    private float alphaAddValue = 0.0f;

    public FadeToUniformColorProcedure(LEDDataBundle _bundle) {
        super(_bundle);
        bundle = _bundle;

        targetColor = bundle.colorPrimary;
        baseColor = bundle.colorSecondary;
        duration = _bundle.duration;
        steps = (int) Math.ceil((duration / (ProgramConfiguration.configuration.frametime / 1000.0f)));
        alphaAddValue = 1 / (float)steps;
    }

    @Override
    public void start() {
        super.start();
    }

    @Override
    public void update() {
        counter++;

        alphaStep += alphaAddValue;
        ColorRGBA outputColor = new ColorRGBA(baseColor.r, baseColor.g, baseColor.b, (int)(alphaStep * 255));
        strip.setAllPixels(outputColor.toRGB(targetColor).toSystemColor());
        if(counter > steps) {
            strip.setAllPixels(baseColor.toSystemColor());
            finishProcedure();
        }
    }
}

File: src/main/java/com/github/iamniklas/liocorekotlin/procedures/FadeToUniformColorProcedure.kt
package com.github.iamniklas.liocorekotlin.procedures

import com.github.iamniklas.liocorekotlin.led.ColorRGB
import com.github.iamniklas.liocorekotlin.led.ColorRGBA
import com.github.iamniklas.liocorekotlin.led.LEDDataBundle

//Fade every pixel from a base color to a target color
class FadeToUniformColorProcedure(_bundle: com.github.iamniklas.liocorekotlin.led.LEDDataBundle) : com.github.iamniklas.liocorekotlin.procedures.Procedure(_bundle) {
    var mBaseColor: com.github.iamniklas.liocorekotlin.led.ColorRGB?
    var mTargetColor: com.github.iamniklas.liocorekotlin.led.ColorRGB?
    var mDuration = 0.0f
    var mCounter = 0
    override var mSteps = 0
    var mAlphaStep = 0.0f
    private var mAlphaAddValue = 0.0f
    override fun start() {}
    public override fun update() {
        mCounter++
        mAlphaStep += mAlphaAddValue
        val outputColor = com.github.iamniklas.liocorekotlin.led.ColorRGBA(
            mBaseColor!!.r,
            mBaseColor!!.g,
            mBaseColor!!.b,
            (mAlphaStep * 255).toInt()
        )
        mStrip!!.setAllPixels(outputColor.toRGB(mTargetColor!!).toSystemColor())
        if (mCounter > mSteps) {
            mStrip!!.setAllPixels(mBaseColor!!.toSystemColor())
            finishProcedure()
        }
    }

    init {
        mTargetColor = _bundle.colorPrimary
        mBaseColor = _bundle.colorSecondary
        mDuration = _bundle.duration!!.toFloat()
        mSteps = Math.ceil((mDuration / (mStrip!!.frametime / 1000.0f)).toDouble()).toInt()
        mAlphaAddValue = 1 / mSteps.toFloat()
    }
}